src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +36,216 @@ case class NoOpFingerprint() {
-  object Directory {
-    // modifiedTime is an Epoch time in milliseconds. (ms since 1970-01-01T00:00:00.000 UTC).
-    case class FileInfo(name: String, size: Long, modifiedTime: Long)
-    object FileInfo {
-      def apply(s: FileStatus): FileInfo =
-        FileInfo(s.getPath.toString, s.getLen, s.getModificationTime)
+
+  /**
+   * Create a Content object from a directory path by recursively listing its leaf files. All
+   * files from the directory tree will be part of the Directory.
+   *
+   * @param path Starting directory path under which the files will be considered part of the
+   *             Directory object.
+   * @param pathFilter Filter for accepting paths. The default filter is picked from spark
+   *                   codebase, which filters out files like _SUCCESS.
+   * @param throwIfNotExists Throws FileNotFoundException if path is not found. Else creates a
+   *                         blank directory tree with no files.
+   * @return Content object with Directory tree starting at root, and containing all leaf files
+   *         from "path" argument.
+   */
+  def fromDirectory(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false): Content =
+    Content(Directory.fromDirectory(path, pathFilter, throwIfNotExists))
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Content = Content(Directory.fromLeafFiles(files))
+}
+
+case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+
+  /**
+   * Create a Directory object from a directory path by recursively listing its leaf files. All
+   * files from the directory tree will be part of the Directory.
+   *
+   * @param path Starting directory path under which the files will be considered part of the
+   *             Directory object.
+   * @param pathFilter Filter for accepting paths. The default filter is picked from spark
+   *                   codebase, which filters out files like _SUCCESS.
+   * @param throwIfNotExists If true, throw FileNotFoundException if path is not found. If set to
+   *                         false, create a blank directory tree with no files.
+   * @return Directory tree starting at root, and containing the files from "path" argument.
+   */
+  def fromDirectory(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false): Directory = {
+    val fs = path.getFileSystem(new Configuration)
+    val leafFiles = listLeafFiles(path, pathFilter, throwIfNotExists, fs)
+
+    if (leafFiles.nonEmpty) {
+      fromLeafFiles(leafFiles)
+    } else {
+      val rootPath = getRoot(path)
```

Comment:
Do you have a test coverage for this `else`?

Reply:
`if (rootPath.equals(path))` could be tricky, but the other `else` should be testable.

Reply:
thanks @imback82 , I cleaned up the old code and added test for empty directories as well as non-existing directories. Both these will lead to the else block.